File: 1513-number-of-substrings-with-only-1s.java
import java.util.*;
import java.util.Map.*;

class NumberOfSubstringsWithOnly1s {
    public int numSub(String s) {
        Map<Long, Long> oneGroupings = new HashMap<>();
        long onesCount = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '0') {
                oneGroupings.put(onesCount, oneGroupings.getOrDefault(onesCount, 0L) + 1);
                onesCount = 0;
                continue;
            }
            onesCount++;
        }
        oneGroupings.put(onesCount, oneGroupings.getOrDefault(onesCount, 0L) + 1);

        int res = 0;
        for (Entry<Long, Long> entry : oneGroupings.entrySet()) {
            long subgroup = entry.getKey(), count = entry.getValue();
            long temp = (long) ((count * (((subgroup) * (subgroup + 1)) / 2)) % (Math.pow(10, 9) + 7)) ;
            res += (int) temp;
            res %= Math.pow(10, 9) + 7;
        }

        return res;
    }
  
    public static void main(String[] args) {

    }
}

File: 1513-number-of-substrings-with-only-1s.py
from collections import defaultdict


class Solution:
    def numSub(self, s: str) -> int:
        '''
        We find the gaps (or groups of ones), count how many are there, differentiated by the length of the substring.
        Next, we calculate how many subgroups are there for each group using math formula (e.g 2 ones = 3, 3 ones = 6, 4 ones = 10, 5 ones = 15)
        Formula is (n (n + 1)) / 2
        '''
        one_groupings = defaultdict(int)
        ones_count = 0
        for c in s:
            if c == '0':
                one_groupings[ones_count] += 1
                ones_count = 0
                continue

            ones_count += 1
        one_groupings[ones_count] += 1

        res = 0
        for subgroup, count in one_groupings.items():
            res += count * (((subgroup) * (subgroup + 1)) // 2)
            res %= (10**9) + 7

        return res


sol = Solution()
print(f'output: {sol.numSub("0110111")}, expected: 9')
print(f'output: {sol.numSub("101")}, expected: 2')
print(f'output: {sol.numSub("111111")}, expected: 21')
